opm/grid/common/SubGridView.hpp

```diff
@@ -0,0 +1,290 @@
+
+public:
+    typedef SubGridViewTraits<GridImp> Traits;
+
+    /** \brief type of the grid */
+    typedef typename Traits::Grid Grid;
+
+    /** \brief type of the index set */
+    typedef typename Traits ::IndexSet IndexSet;
+
+    /** \brief type of the intersection */
+    typedef typename Traits ::Intersection Intersection;
+
+    /** \brief type of the intersection iterator */
+    typedef typename Traits ::IntersectionIterator IntersectionIterator;
+
+    /** \brief type of the collective communication */
+    typedef typename Traits ::CollectiveCommunication CollectiveCommunication;
+
+    /** \brief Codim Structure */
+    template <int cd>
+    struct Codim : public Traits ::template Codim<cd> {
+        using Entity = typename Traits::template Codim<cd>::Entity;
+        class Iterator
+        {
+        public:
+            Iterator(const SubGridView& view, std::size_t index)
+                : view_(&view)
+                , index_(index)
+            {
+            }
+            const Entity& operator*() const
+            {
+                entity_ = this->view_->get(index_);
+                return entity_;
+            }
+            const Entity* operator->() const
+            {
+                entity_ = this->view_->get(index_);
+                return &entity_;
+            }
+            Iterator operator++()
+            {
+                ++index_;
+                return *this;
+            }
+            Iterator operator++(int)
+            {
+                Iterator copy(*this);
+                ++index_;
+                return copy;
+            }
+            bool operator==(const Iterator& other) const
+            {
+                assert(view_ == other.view_);
+                return index_ == other.index_;
+            }
+            bool operator!=(const Iterator& other) const
+            {
+                assert(view_ == other.view_);
+                return index_ != other.index_;
+            }
+        private:
+            const SubGridView* view_;
+            std::size_t index_;
+            mutable Entity entity_; // This may be low-performing for grids with large Entity objects.
+        };
+    };
+
+    enum { conforming = Traits::conforming };
+    enum { dimension = GridImp::dimension };
+
+public:
+    /// Construct a view of the codim 0 entities enumerated in the indices vector.
+    ///
+    /// The indices vector is assumed to be sorted in ascending order.
+    SubGridView(const Grid& grid,
+                const std::vector<int>& indices)
+        : grid_(&grid)
+    {
+        assert(std::is_sorted(indices.begin(), indices.end()));
+        subset_.resize(indices.size());
+        const auto beg = grid.template leafbegin<0>();
+        auto it = beg;
+        int previous = 0;
+        for (size_t c = 0; c < indices.size(); ++c) {
+            using EntitySeed = typename Codim<0>::Entity::EntitySeed;
+            std::advance(it, indices[c] - previous);
```

Comment:
It is not that expensive, as it only steps the difference between the previous and next. But yes, it will make one iteration sweep of the entire grid to create one SubGridView. If all you need is a single such view, this constructor is as fast as it gets, the other constructor will need to get the seeds from a similar grid iteration sweep. If you want to construct many sub views, you should try to make all with a single sweep, and for that you need the other constructor. The question is if it is dangerous to keep this one? I am on the side of "not very".

Reply:
> If all you need is a single such view, this constructor is as fast as it gets, the other constructor will need to get the seeds from a similar grid iteration sweep.

True, but that's not really an interesting use case for a subgrid view.

> The question is if it is dangerous to keep this one? I am on the side of "not very".

It Depends[tm].  The mere existence of this constructor will, at least in some small way, encourage its use.  I'd rather that there be a single, canonical constructor and then applications can create free functions to form objects from other types if needed.